### PR TITLE
add helper method for service error unmarshal failures

### PR DESCRIPTION
**Pull request by catalinaperalta**
- [ ] The purpose of this PR is explained in this or a referenced issue.
- [ ] The PR does not update generated files.
   - These files are managed by the codegen framework at [Azure/autorest.go][].
- [ ] Tests are included and/or updated for code changes.
- [ ] Updates to [CHANGELOG.md][] are included.
- [ ] MIT license headers are included in each file.
 
[Azure/autorest.go]: https://github.com/Azure/autorest.go
[CHANGELOG.md]: https://github.com/Azure/azure-sdk-for-go/blob/master/CHANGELOG.md

### CODE REVIEW

**Comment by catalinaperalta** [comment]
I forced an unmarshal error to get the result of calling response.UnmarshalError(). Real world example: 
```
raw service error: {"code":"LibraryDoesNotExistInWorkspace","message":"The LibraryArtifact test does not exist."}
unmarshalling type azartifacts.CloudErrorAutoGenerated: json: Unmarshal(non-pointer azartifacts.CloudErrorAutoGenerated)
```

**Comment by catalinaperalta** [comment]
generated error handler code should look like:
```
// getHandleError handles the Get error response.
func (client *LibraryClient) getHandleError(resp *azcore.Response) error {
	var err CloudErrorAutoGenerated
	if err := resp.UnmarshalAsJSON(err); err != nil {
		return resp.UnmarshalError(err)
	}
	return azcore.NewResponseError(&err, resp.Response)
}
```

File: sdk/azcore/response_test.go [hunk 1]
```diff
@@ -7,6 +7,7 @@ package azcore
```
Comment (ArcturusZhang) [inline sdk/azcore/response_test.go:7]: I am curious that should we unify the package name of the test packages? I am seeing one of the test package `azcore_test` here: https://github.com/Azure/azure-sdk-for-go/blob/master/sdk/azcore/example_test.go

Comment (catalinaperalta) [reply sdk/azcore/response_test.go:7]: Good point! Makes sense to make these also azcore_test. I can update this in a separate PR though.

Comment (jhendrixMSFT) [reply sdk/azcore/response_test.go:7]: So I believe the name `azcore_test` is intentional in this case.  When you use the package name `<package>_test` it means you don't have access to unexported content within `<package>`.  This is useful when writing examples for two reasons.

- the examples will have to use the fully qualified name, e.g. `azcore.Poller` which is nice for copy/paste
- it ensures you don't have access to unexported content

Comment (catalinaperalta) [reply sdk/azcore/response_test.go:7]: Gotcha, will leave as is.

**Comment by catalinaperalta** [comment]
> generated error handler code should look like:
> 
> ```
> // getHandleError handles the Get error response.
> func (client *LibraryClient) getHandleError(resp *azcore.Response) error {
> 	var err CloudErrorAutoGenerated
> 	if err := resp.UnmarshalAsJSON(err); err != nil {
> 		return azcore.NewResponseError(resp.UnmarshalError(err), resp.Response)
> 	}
> 	return azcore.NewResponseError(&err, resp.Response)
> }
> ```
> 
> EDIT: actually changed this to also return the response error type, we can discuss this on the generator side

@jhendrixMSFT let me know if you think this is a good idea, I will wait to open the autorest PR until I hear your opinion on this.

File: sdk/azcore/response.go [hunk 2]
```diff
@@ -129,6 +129,16 @@ func (r *Response) UnmarshalAsXML(v interface{}) error {
...
 }
 
+// UnmarshalError will unmarshal the received service error payload along with the corresponding unmarshal error.
+// NOTE: this method is only meant to be called when handling a service error.
+func (r *Response) UnmarshalError(unmarshalErr error) error {
+	payload, err := r.payload()
+	if err != nil {
+		return err
+	}
+	return fmt.Errorf("raw service error: %s\n%w", payload, unmarshalErr)
```
Comment (jhendrixMSFT) [inline sdk/azcore/response.go:139]: I wouldn't wrap `unmarshalErr` here since we don't need/want callers to test for explicit unmarshall errors.

Comment (catalinaperalta) [reply sdk/azcore/response.go:139]: Didnt we also want include the unmarshal error for ourselves?

File: sdk/azcore/response.go [hunk 3]
```diff
@@ -129,6 +129,16 @@ func (r *Response) UnmarshalAsXML(v interface{}) error {
 	return err
 }
 
+// UnmarshalError will unmarshal the received service error payload along with the corresponding unmarshal error.
+// NOTE: this method is only meant to be called when handling a service error.
+func (r *Response) UnmarshalError(unmarshalErr error) error {
+	payload, err := r.payload()
```
Comment (jhendrixMSFT) [inline sdk/azcore/response.go:135]: I believe it's possible (although unlikely) that `payload` could be empty.  If this is the case, we have two options.

- add some default message "missing body" or whatever
- just return `unmarshalErr`

I _think_ that the JSON/XML unmarshallers will fail with a message indicating the first case ("unexpected end of blah blah").  If that's the case then let's do the second one.

File: sdk/azcore/response.go (hunk 2 above)
Comment (jhendrixMSFT) [reply sdk/azcore/response.go:139]: We want to include it but not wrap it, so just a `%s` should do the trick.

**Comment by jhendrixMSFT** [comment]
So this will be used like so right?

```go
func (client *LROsClient) delete202NoRetry204HandleError(resp *azcore.Response) error {
	var err CloudError
	if err := resp.UnmarshalAsJSON(&err); err != nil {
		return azcore.UnmarshalError(err)
	}
	return azcore.NewResponseError(&err, resp.Response)
}
```

File: sdk/azcore/response.go (hunk 3 above)
Comment (catalinaperalta) [reply sdk/azcore/response.go:135]: I will check this and simply return the unmarshalErr if the other unmarshallers have a good message.

**Comment by catalinaperalta** [comment]
> So this will be used like so right?
> 
> ```go
> func (client *LROsClient) delete202NoRetry204HandleError(resp *azcore.Response) error {
> 	var err CloudError
> 	if err := resp.UnmarshalAsJSON(&err); err != nil {
> 		return azcore.UnmarshalError(err)
> 	}
> 	return azcore.NewResponseError(&err, resp.Response)
> }
> ```

Well I initially had this, but then thought that we could wrap the error from azcore.UnmarshalError() in an azcore.NewResponseError() so that users can access the service error themselves and check for status codes, etc. What do you think?

**Comment by jhendrixMSFT** [comment]
Yes you're correct, so it would be `returnazcore.NewResponseError( azcore.UnmarshalError(err), resp.Response)`

File: sdk/azcore/response.go (hunk 3 above)
Comment (catalinaperalta) [reply sdk/azcore/response.go:135]: It doesnt seem that the other unmarshallers add a message if the payload is empty here's one of them as reference:
```
func (r *Response) UnmarshalAsJSON(v interface{}) error {
	payload, err := r.payload()
	if err != nil {
		return err
	}
	// TODO: verify early exit is correct
	if len(payload) == 0 {
		return nil
	}
	err = r.removeBOM()
	if err != nil {
		return err
	}
	err = json.Unmarshal(payload, v)
	if err != nil {
		err = fmt.Errorf("unmarshalling type %T: %s", v, err)
	}
	return err
}
```

Comment (catalinaperalta) [reply sdk/azcore/response.go:135]: I think it's still a good idea to just include the unmarshalErr even if the payload is empty, since the err we're sending is populated with the error from json.Ummarshal()

File: sdk/azcore/response.go (hunk 2 above)
Comment (catalinaperalta) [reply sdk/azcore/response.go]: Done

File: sdk/azcore/response.go (hunk 3 above)
Comment (catalinaperalta) [reply sdk/azcore/response.go:135]: I added empty error body to the message. Done?

Comment (jhendrixMSFT) [reply sdk/azcore/response.go:135]: Oh yes I forgot we have that empty body check.  We should follow up on that separately to see if that's really what we should be doing (I forget why I added that, might have been early on prototyping).

File: sdk/azcore/response.go [hunk 4]
```diff
@@ -129,6 +129,19 @@ func (r *Response) UnmarshalAsXML(v interface{}) error {
 	return err
 }
 
+// UnmarshalError will unmarshal the received service error payload along with the corresponding unmarshal error.
```
Comment (jhendrixMSFT) [inline sdk/azcore/response.go:132]: Nit, I would reword this slightly this doesn't unmarshall stuff.

**Review by jhendrixMSFT — APPROVED** [review]
Small rewording of doc comment.
